sys/net/application_layer/nanocoap/sock.c

```diff
@@ -315,24 +311,38 @@ ssize_t nanocoap_sock_request_cb(nanocoap_sock_t *sock, coap_pkt_t *pkt,
                     continue;
                 }
                 /* fall-through */
+            case COAP_TYPE_RST:
             case COAP_TYPE_CON:
             case COAP_TYPE_NON:
                 state = STATE_RESPONSE_OK;
-                if (coap_get_type(pkt) == COAP_TYPE_CON) {
-                    /* send ACK */
-                    _send_ack(sock, pkt);
+                if (coap_get_type(pkt) == COAP_TYPE_RST) {
+                     /* think about whether cb should be called on RST as well */
+                    res = -EBADMSG;
                 }
-                /* call user callback */
-                if (cb) {
-                    res = cb(arg, pkt);
-                } else {
-                    res = _get_error(pkt);
+                else {
+                    if (coap_get_type(pkt) == COAP_TYPE_CON) {
+                        _send_ack(sock, pkt);
+                    }
+                    if (cb) {
+                        res = cb(arg, pkt);
+                    }
+                    else {
+                        res = _get_error(pkt);
+                    }
```

Comment:
```suggestion
              if (coap_get_type(pkt) == COAP_TYPE_CON) {
                  _send_ack(sock, pkt);
              }
              if (cb) {
                  res = cb(arg, pkt);
              }
              else {
                  res = _get_error(pkt);
```

together with the above suggestion